### ISSUE TITLE
TileLayer with BitmapLayer

### ISSUE THREAD

**Issue by mbarrenechea** [uber/deck.gl#2713]
Hi,

I tried to use a combination between TileLayer and BitmapLayer because I want to parse each pixel of a Raster. 

The thing is that I managed to show it on the map applying my decode function => https://github.com/mbarrenechea/mapbox-sandbox/blob/master/src/components/bitmap-layer/bitmap-layer-fragment.js#L37

The problem is that my raster images seem to be blurred. Even if I don't apply my decode function, the result is the same, so I discarded that the problem is on my decode function.

![image](https://user-images.githubusercontent.com/8928965/53229799-7ac82c80-3685-11e9-96c6-7ce670f7dc75.png).

The desired result should be this one that you can see [here](https://www.globalforestwatch.org/map?map=eyJkYXRhc2V0cyI6W3siZGF0YXNldCI6ImZkYzhkYzFiLTI3MjgtNGE3OS1iMjNmLWIwOTQ4NTA1MmI4ZCIsImxheWVycyI6WyI2ZjY3OThlNi0zOWVjLTQxNjMtOTc5ZS0xODJhNzRjYTY1ZWUiLCJjNWQxZTAxMC0zODNhLTQ3MTMtOWFhYS00NGY3MjhjMDU3MWMiXSwib3BhY2l0eSI6MSwidmlzaWJpbGl0eSI6dHJ1ZX0seyJkYXRhc2V0IjoiODk3ZWNjNzYtMjMwOC00YzUxLWFlYjMtNDk1ZGUwYmRjYTc5IiwibGF5ZXJzIjpbImMzMDc1YzVhLTU1NjctNGIwOS1iYzBkLTk2ZWQxNjczZjhiNiJdLCJvcGFjaXR5IjoxLCJ2aXNpYmlsaXR5Ijp0cnVlfV0sInpvb20iOjYsImNlbnRlciI6eyJsYXQiOjQ3LjMzMTM3NzE1Nzc5ODI0NCwibG5nIjotODcuMjMxNDQ1MzEyNTAwMDF9LCJjYW5Cb3VuZCI6ZmFsc2UsImJib3giOm51bGx9) 
In this case, we are using leaflet + canvas to decode each of my raster images

https://github.com/mbarrenechea/mapbox-sandbox

What do you think? Do you think is possible to do that? Thank you in advance

**Comment by jianhuang01** [comment]
the blur is from texture filtering, you can try to set the texture parameters as 
const TEX_PARAMETER = {
  [GL.TEXTURE_MIN_FILTER]: GL.NEAREST,
  [GL.TEXTURE_MAG_FILTER]: GL.NEAREST,
  [GL.TEXTURE_WRAP_S]: GL.CLAMP_TO_EDGE,
  [GL.TEXTURE_WRAP_T]: GL.CLAMP_TO_EDGE
};
 and disable mipmaps
Here is the [Texture2D doc](https://github.com/uber/luma.gl/blob/master/docs/api-reference/webgl/texture-2d.md)

**Comment by edbrett** [comment]
hi @jianhuang01 thanks, for responding to this. I have been working on this with @mbarrenechea.

I have followed your response and looked through the docs for the Texture2D but no options seem to prevent the blurring on the tiles images. I have found that applying the parameters you specified seems to prevent the blurring of the textures when at zoom levels beyond the max zoom (in this case 11). We have tiles up to zoom 11. I have attached a gif of starting at zoom 12 and zooming out to < 11 to show you the difference. I have also attached an image of what the tiles look like without being passed through out custom decode as that shows better the individual textures with the added darkening and blurring around the edges.

Zooming:
![2019-02-26 15 04 17](https://user-images.githubusercontent.com/20288774/53418760-d0c80780-39d8-11e9-9443-be2f0d33b490.gif)

The blurring effect at high zoom:
![screenshot 2019-02-26 at 15 11 00](https://user-images.githubusercontent.com/20288774/53418764-d3c2f800-39d8-11e9-804c-73fccf7d31ae.png)

Any thoughts on why would be greatly appreciated!

**Comment by tsherif** [comment]
Hi @edbrett. It looks like mipmapping is still enabled there. Do you have a live version of your app available somewhere? Might be able to narrow this down if I run it through some inspection tools.

**Comment by edbrett** [comment]
Hi @tsherif, thanks for getting back. I have just pushed on master the latest of what we are working on with the above suggested changes here: https://github.com/mbarrenechea/mapbox-sandbox.

I don't have anything deployed but this should be easy to clone and run.

**Comment by jianhuang01** [comment]
@edbrett Played your app for a while, downloaded two pngs from the data source, opened by macbook Preview app and zoom in.
[Image from zoom 12](https://storage.googleapis.com/wri-public/Hansen17/tiles/hansen_world/v1/tc30/12/1077/1653.png)
![image](https://user-images.githubusercontent.com/15093546/53448063-665c9a80-39cb-11e9-8063-3f6506d2fffd.png)
[Image from zoom 10](https://storage.googleapis.com/wri-public/Hansen17/tiles/hansen_world/v1/tc30/10/206/317.png)
![image](https://user-images.githubusercontent.com/15093546/53448143-9015c180-39cb-11e9-947c-dbb8f2f295b9.png)
As you see above, the raw png from zoom 10 already has blur blended.

By the way, I can't run the app by simply clone your [repo](https://github.com/mbarrenechea/mapbox-sandbox), because of wrong file paths like **import Map from 'components/map'**

**Comment by mbarrenechea** [comment]
If you want to run the app, remember to create a file named .env in the root with the .env.sample content. It's on the readme.

Thank you so much for helping us, tomorrow we will try another source for the tiles

**Comment by edbrett** [comment]
@jianhuang01 I have been looking at the source tiles also and comparing them. You are correct from the example you gave that there is some alpha on the source tiles below z12, but the issue we found with this was something else. I found that if you decode the tiles and return a vertex with alpha set to 1, then there is no blurring effect at any level, but when you apply an alpha, therefore meaning that pixels that are next to each other have a different colour, there is some artefacts shown in the blending of those colours. To demonstrate this I used a different layer that has better colour contrasts.

Layer source tile from source:
![2c4fde29-1627-40eb-96b5-a9e388f7c7b7_10_363_606_tile_269c842b01900a59075f395b44be3bb0](https://user-images.githubusercontent.com/20288774/53572349-5380cc00-3b6b-11e9-8d5a-67911ceabae1.png)

GL decoded layer without alpha (set alpha band to 1 on all pixels):
![screenshot 2019-02-28 at 15 21 27](https://user-images.githubusercontent.com/20288774/53572910-a444f480-3b6c-11e9-8e83-b0b87a2331e8.png)

GL decoded layer with alpha applied (variable alpha):
![screenshot 2019-02-28 at 15 21 43](https://user-images.githubusercontent.com/20288774/53572919-a8711200-3b6c-11e9-8975-e00c0be0f01f.png)

We played around with the settings for the textures for a while but couldn't find anything that improved the blending of the opaque pixels. However, one thing we did notice inside the TileLayer we inherited from the experimental layers is that we might be accessing the incorrect tiles for the given zoom. We noticed in the raster layer in mapbox itself when taking a raster source, they fetched the tiles at the .ceil() of the current zoom float, but in the TileLayer we are getting the .floor(). We changed this and made some modifications to the viewport calculations and it seems to improve the resolution of the tiles in view significantly. Example:
![screenshot 2019-02-28 at 15 22 07](https://user-images.githubusercontent.com/20288774/53572932-ad35c600-3b6c-11e9-80a9-03fe2d0b00bd.png)

After also fixing our decode function within our shader, we managed to get the original layer looking as we wanted. Here it is with an animation applied:
![2019-02-28 12 09 32](https://user-images.githubusercontent.com/20288774/53573035-ebcb8080-3b6c-11e9-98cc-87884f3d110a.gif)

In summary we feel we have found a solution by changing the tiles that are fetched for a given zoom, but there does seem to be some noticeable colouring when blending tiles with variable alpha between adjacent pixels. If this can be improved, or you have any suggestions on how to sharpen blending of alpha on pixel we would be very grateful!? Otherwise we are happy with what we have.

**Comment by ibgreen** [comment]
Sounds like you may want to try changing blend settings. Using the `parameters` prop on the layer you can control any WebGL settings including all blend settings, some examples:

```
        parameters: {
          blendEquation: GL.MAX
        }
```

```
   parameters: {
      blendFunc: [gl.SRC_ALPHA, gl.ONE_MINUS_SRC_ALPHA, gl.ONE, gl.ONE_MINUS_SRC_ALPHA],
      blendEquation: gl.FUNC_ADD
   }
```

etc.

For more information see luma.gl docs and any WebGL article about blending. I wish we had a good guide, unfortunately, blend settings are not exactly trivial to understand/explain, but often one can get significant improvements in visuals by using them, so worth playing around.

**Comment by jianhuang01** [comment]
@edbrett Glad to see you have progress, actually I don't know how you decode the data, that's not our API, so can't help on that area.

**Comment by HeiYanjing** [comment]
@mbarrenechea  have you ever load a mapbox  tile layer?  
#3329  I tried load a building layer published by geoserver ,could help have a look?

**Comment by mbarrenechea** [comment]
@HeiYanjing no, our source of data is a raster tile layer. I'll take a look at your code, but it doesn't seem to be related

**Comment by HeiYanjing** [comment]
@mbarrenechea  thank you!

**Comment by tgorkin** [comment]
Closing issue as I believe @ibgreen's last response answers the remaining question about how to control alpha blending on a layer using the `parameters` prop.
